README.md
![FloatingTextField: A Customizable Text Field](https://raw.githubusercontent.com/space-code/floating-text-field/dev/Resources/floating-text-field.png)

<h1 align="center" style="margin-top: 0px;">floating-text-field</h1>

<p align="center">
<a href="https://github.com/space-code/floating-text-field/blob/main/LICENSE"><img alt="License" src="https://img.shields.io/github/license/space-code/floating-text-field?style=flat"></a> 
<a href="https://developer.apple.com/"><img alt="Platform" src="https://img.shields.io/badge/platform-ios-%23989898"/></a> 
<a href="https://developer.apple.com/swift"><img alt="5.7" src="https://img.shields.io/badge/language-Swift5.7-orange.svg"/></a>
<a href="https://github.com/space-code/floating-text-field"><img alt="CI" src="https://github.com/space-code/floating-text-field/actions/workflows/ci.yml/badge.svg?branch=main"></a>
<a href="https://github.com/apple/swift-package-manager" alt="floating-text-field on Swift Package Manager" title="floating-text-field on Swift Package Manager"><img src="https://img.shields.io/badge/Swift%20Package%20Manager-compatible-brightgreen.svg" /></a>
<a href="https://codecov.io/gh/space-code/floating-text-field"><img alt="CodeCov" src="https://codecov.io/gh/space-code/floating-text-field/graph/badge.svg?token=lA8TWRHGkf"></a>
</p>

## Description
`floating-text-field` is a customizable text field.

- [Visual Example](#visual-example)
- [Usage](#usage)
- [Requirements](#requirements)
- [Installation](#installation)
- [Communication](#communication)
- [Contributing](#contributing)
- [Author](#author)
- [License](#license)

## Visual Example

The visual example of the `FloatingTextField`:

![Example](Resources/examples/example.gif)

## Usage

### Basic Usage

For basic usage, simply import the `FloatingTextField` package and set the text field height as follows:

```swift
import FloatingTextField

struct ContentView: View {
    @State private var username: String = ""

    var body: some View {
        FloatingTextFieldView($username, placeholder: "username")
            .frame(height: 60.0)
    }
}
```

### Custom Fonts and Colors

The `FloatingTextField` provides an opportunity to customize text font, text color, placeholder font, placeholder color, and more.

```swift
import FloatingTextField

struct ContentView: View {
    @State private var username: String = ""

    var body: some View {
        FloatingTextFieldView($username, placeholder: "username")
            .font(Font.system(size: 17.0))
            .placeholderFont(Font.system(size: 14.0))
            .textColor(.black)
            .placeholderColor(.gray)
            .frame(height: 60.0)
    }
}
```

### Custom Styles

You can create a text field style in two simple steps.

1. Define a custom text field style as follows:

```swift
import FloatingTextField

struct CustomTextFieldStyle: FloatingTextFieldStyle {
    func body(content: FloatingTextField) -> FloatingTextField {
        content
            .font(Font.system(size: 17.0))
            .placeholderFont(Font.system(size: 14.0))
            .cornerRadius(12)
            .placeholderBottomPadding(4.0)
            .textColor(.black)
            .placeholderColor(.gray)
            .borderWidth(1)
            .borderColor(.black)
    }
}
```

2. Apply this style using `textFieldStyle(_:)`:

```swift
struct ContentView: View {
    @State private var username: String = ""

    var body: some View {
        FloatingTextFieldView($username, placeholder: "username")
            .textFieldStyle(style: CustomTextFieldStyle())
            .frame(height: 60.0)
    }
}
```

### Secure Text Entry

The `FloatingTextField` can conceal sensitive data such as passwords. You can hide it by using the `isSecureTextEntry(_:)` modifier.

```swift
import FloatingTextField

struct ContentView: View {
    @State private var password: String = ""
    @State private var isPasswordHidden: Bool = true

    var body: some View {
        FloatingTextField($password, placeholder: "Password")
            .textFieldStyle(style: CustomTextFieldStyle())
            .isSecureTextEntry(isPasswordHidden)
            .leftView {
                Image(systemName: "lock")
                    .foregroundColor(.gray)
            }
            .rightView {
                Button(
                    action: { isPasswordHidden.toggle() },
                    label: {
                        Image(systemName: "eye.fill")
                            .foregroundColor(.gray)
                    }
                )
            }
            .frame(height: 60.0)
    }
}
```

## Requirements

- iOS 15.0+
- Xcode 14.0
- Swift 5.7

## Installation
### Swift Package Manager

The [Swift Package Manager](https://swift.org/package-manager/) is a tool for automating the distribution of Swift code and is integrated into the `swift` compiler. It is in early development, but `floating-text-field` does support its use on supported platforms.

Once you have your Swift package set up, adding `floating-text-field` as a dependency is as easy as adding it to the `dependencies` value of your `Package.swift`.

```swift
dependencies: [
    .package(url: "https://github.com/space-code/floating-text-field.git", .upToNextMajor(from: "1.0.0"))
]
```

## Communication
- If you **found a bug**, open an issue.
- If you **have a feature request**, open an issue.
- If you **want to contribute**, submit a pull request.

## Contributing
Bootstrapping development environment

```
make bootstrap
```

Please feel free to help out with this project! If you see something that could be made better or want a new feature, open up an issue or send a Pull Request!

## Author
Nikita Vasilev, nv3212@gmail.com

## License
floating-text-field is available under the MIT license. See the LICENSE file for more info.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

#### 1.x Releases
- `1.0.x` Releases - [1.0.0](#100)

## [1.0.0](https://github.com/space-code/floating-text-field/releases/tag/1.0.0)
Released on 2023-10-18.

#### Added
- Initial release of log.
  - Added by [Nikita Vasilev](https://github.com/nik3212).

Tests/FloatingTextFieldTests/FloatingTextFieldTests.swift
//
// floating-text-field
// Copyright © 2023 Space Code. All rights reserved.
//

import XCTest

final class FloatingTextFieldTests: XCTestCase {}


Tests/FloatingTextFieldTests/UnitTests/FloatingTextFieldTests.swift
//
// floating-text-field
// Copyright © 2023 Space Code. All rights reserved.
//

@testable import FloatingTextField
import SwiftUI
import XCTest

// MARK: - FloatingTextFieldTests

final class FloatingTextFieldTests: XCTestCase {
    // MARK: Properties

    private var sut: FloatingTextField!

    // MARK: XCTestCase

    override func setUp() {
        super.setUp()
        sut = FloatingTextField(
            .constant(.text),
            placeholder: .placeholder
        )
    }

    override func tearDown() {
        sut = nil
        super.tearDown()
    }

    // MARK: Tests

    func test_thatFloatingTextFieldChangesForegroundColor_whenTitleColorDidSet() {
        // when
        _ = sut.titleColor(.red)

        // then
        XCTAssertEqual(sut.configuration.titleColor, .red)
    }

    func test_thatFloatingTextFieldChangesSecureEntry_whenSecureEntryDidSet() {
        // when
        _ = sut.secureField(true)

        // then
        XCTAssertTrue(sut.configuration.isSecureEntry)
    }

    func test_thatFloatingTextFieldChangesPlaceholderColor_whenPlaceholderColorDidSet() {
        // when
        _ = sut.placeholderColor(.red)

        // then
        XCTAssertEqual(sut.configuration.placeholderColor, .red)
    }

    func test_thatFloatingTextFieldChangesPlaceholderFont_whenPlaceholderFontDidSet() {
        // given
        let font = Font.system(size: 17.0)

        // when
        _ = sut.placeholderFont(font)

        // then
        XCTAssertEqual(sut.configuration.placeholderFont, font)
    }

    func test_thatFloatingTextFieldChangesFont_whenFontDidSet() {
        // given
        let font = Font.system(size: 17.0)

        // when
        _ = sut.textFont(font)

        // then
        XCTAssertEqual(sut.configuration.font, font)
    }

    func test_thatFloatingTextFieldChangesBorderWidth_whenBorderWidthDidSet() {
        // given
        let borderWidth: CGFloat = 1

        // when
        _ = sut.borderWidth(borderWidth)

        // then
        XCTAssertEqual(sut.configuration.borderWidth, borderWidth)
    }

    func test_thatFloatingTextFieldChangesBorderColor_whenBorderColorDidSet() {
        // when
        _ = sut.borderColor(.red)

        // then
        XCTAssertEqual(sut.configuration.borderColor, .red)
    }

    func test_thatFloatingTextFieldChangesFocusedBorderColor_whenFocusedBorderColorDidSet() {
        // when
        _ = sut.focusedBorderColor(.red)

        // then
        XCTAssertEqual(sut.configuration.focusedBorderColor, .red)
    }

    func test_thatFloatingTextFieldChangesBorderCornerRadius_whenFocusedBorderCornerRadiusDidSet() {
        // given
        let cornerRadius: CGFloat = 12

        // when
        _ = sut.borderCornerRadius(cornerRadius)

        // then
        XCTAssertEqual(sut.configuration.borderCornerRadius, cornerRadius)
    }

    func test_thatFloatingTextFieldChangesPlaceholderBottomPadding_whenPlaceholderBottomPaddingDidSet() {
        // given
        let padding: CGFloat = 12

        // when
        _ = sut.placeholderBottomPadding(padding)

        // then
        XCTAssertEqual(sut.configuration.placeholderBottomPadding, padding)
    }
}

// MARK: - Constants

private extension String {
    static let text = "text"
    static let placeholder = "placeholder"
}


Tests/FloatingTextFieldTests/SnapshotTests/FloatingTextFieldSnapshotTests.swift
//
// floating-text-field
// Copyright © 2023 Space Code. All rights reserved.
//

@testable import FloatingTextField
import SnapshotTesting
import SwiftUI
import XCTest

// MARK: - FloatingTextFieldSnapshotTests

final class FloatingTextFieldSnapshotTests: XCTestCase {
    // MARK: Tests

    func test_thatFloatingTextFieldShowsOnlyPlaceholder_whenTextIsEmpty() {
        assertSnapshots(
            of: FloatingTextField(.constant(.empty), placeholder: .placeholder),
            as: [.image(layout: .default, traits: .light)]
        )
    }

    func test_thatFloatingTextFieldShowsCompactedPlaceholder_whenTextIsNotEmpty() {
        assertSnapshots(
            of: FloatingTextField(.constant(.text), placeholder: .placeholder),
            as: [.image(layout: .default, traits: .light)]
        )
    }

    func test_thatFloatingTextFieldShowsSecureField() {
        assertSnapshots(
            of: FloatingTextField(.constant(.text), placeholder: .placeholder)
                .isSecureTextEntry(true),
            as: [.image(layout: .default, traits: .light)]
        )
    }

    func test_thatFloatingTextFieldShowsCustomizedView() {
        assertSnapshots(
            of: FloatingTextField(.constant(.text), placeholder: .placeholder)
                .font(Font.system(size: 17.0))
                .borderWidth(2.0)
                .borderColor(.black)
                .focusedBorderColor(.black)
                .textColor(Color.black)
                .placeholderFont(Font.system(size: 14.0))
                .placeholderColor(.gray)
                .cornerRadius(12.0)
                .placeholderBottomPadding(.zero)
                .disabled(false)
                .leftView { Image(systemName: "mail.fill").foregroundColor(.gray) }
                .rightView { Image(systemName: "eye.fill").foregroundColor(.gray) },
            as: [.image(layout: .default, traits: .light)]
        )
    }

    func test_thatFloatingTextFieldShowsCustomizedView_whenCustomStyleDidSet() {
        assertSnapshots(
            of: FloatingTextField(.constant(.text), placeholder: .placeholder)
                .textFieldStyle(style: TextFieldStyle()),
            as: [.image(layout: .default, traits: .light)]
        )
    }
}

// MARK: FloatingTextFieldSnapshotTests.TextFieldStyle

private extension FloatingTextFieldSnapshotTests {
    struct TextFieldStyle: FloatingTextFieldStyle {
        func body(content: FloatingTextField) -> FloatingTextField {
            content
                .font(Font.system(size: 17.0))
                .borderWidth(1)
                .borderColor(.gray)
                .cornerRadius(8.0)
        }
    }
}

// MARK: - Constants

private extension String {
    static let empty = ""
    static let text = "name@gmail.com"
    static let placeholder = "Username"
}

private extension SwiftUISnapshotLayout {
    static let `default` = SwiftUISnapshotLayout.fixed(width: 375, height: 60)
}

private extension UITraitCollection {
    static let light = UITraitCollection(userInterfaceStyle: .light)
}


Package.swift
// swift-tools-version: 5.7
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "FloatingTextField",
    platforms: [
        .iOS(.v15),
    ],
    products: [
        .library(name: "FloatingTextField", targets: ["FloatingTextField"]),
    ],
    dependencies: [
        .package(
            url: "https://github.com/pointfreeco/swift-snapshot-testing",
            from: "1.12.0"
        ),
    ],
    targets: [
        .target(name: "FloatingTextField", dependencies: []),
        .testTarget(
            name: "FloatingTextFieldTests",
            dependencies: [
                "FloatingTextField",
                .product(name: "SnapshotTesting", package: "swift-snapshot-testing"),
            ]
        ),
    ]
)


Sources/FloatingTextField/Classes/UI/FloatingTextField+Configuration.swift
//
// floating-text-field
// Copyright © 2023 Space Code. All rights reserved.
//

import SwiftUI

// MARK: - Extensions

public extension FloatingTextField {
    func placeholderColor(_ color: Color) -> FloatingTextField {
        configuration.placeholderColor = color
        return self
    }

    func placeholderFont(_ font: Font) -> FloatingTextField {
        configuration.placeholderFont = font
        return self
    }

    func font(_ font: Font) -> FloatingTextField {
        configuration.font = font
        return self
    }

    func borderColor(_ color: Color) -> FloatingTextField {
        configuration.borderColor = color
        return self
    }

    func borderWidth(_ lineWidth: CGFloat) -> FloatingTextField {
        configuration.borderWidth = lineWidth
        return self
    }

    func cornerRadius(_ cornerRadius: CGFloat) -> FloatingTextField {
        configuration.cornerRadius = cornerRadius
        return self
    }

    func placeholderBottomPadding(_ padding: CGFloat) -> FloatingTextField {
        configuration.placeholderBottomPadding = padding
        return self
    }

    func disabled(_ disabled: Bool) -> FloatingTextField {
        configuration.disabled = disabled
        return self
    }

    func textColor(_ color: Color) -> FloatingTextField {
        configuration.titleColor = color
        return self
    }

    func isSecureTextEntry(_ isSecureTextEntry: Bool) -> FloatingTextField {
        configuration.isSecureTextEntry = isSecureTextEntry
        return self
    }

    func focusedBorderColor(_ borderColor: Color) -> FloatingTextField {
        configuration.focusedBorderColor = borderColor
        return self
    }

    func leftView<V: View>(@ViewBuilder _ view: @escaping () -> V) -> FloatingTextField {
        configuration.leftView = AnyView(view())
        return self
    }

    func rightView<V: View>(@ViewBuilder _ view: @escaping () -> V) -> FloatingTextField {
        configuration.rightView = AnyView(view())
        return self
    }
}


Sources/FloatingTextField/Classes/Models/FloatingTextFieldConfiguration.swift
//
// floating-text-field
// Copyright © 2023 Space Code. All rights reserved.
//

import Foundation
import SwiftUI

/// A configuration class for defining the appearance and behavior of a `FloatingTextField`.
final class FloatingTextFieldConfiguration: ObservableObject {
    // MARK: Properties

    /// Determines whether the text input is treated as a secure entry (e.g., for passwords).
    @Published var isSecureTextEntry: Bool

    /// The color of the title when the text field is in its default state.
    @Published var titleColor: Color

    /// The color of the placeholder text.
    @Published var placeholderColor: Color

    /// The font used for the placeholder text.
    @Published var placeholderFont: Font

    /// The font used for the text input.
    @Published var font: Font

    /// The width of the text field's border.
    @Published var borderWidth: CGFloat

    /// The color of the text field's border.
    @Published var borderColor: Color

    /// The color of the text field's border when it is focused.
    @Published var focusedBorderColor: Color

    /// The corner radius of the text field's border.
    @Published var cornerRadius: CGFloat

    /// The amount of padding beneath the placeholder text.
    @Published var placeholderBottomPadding: CGFloat

    /// A flag to disable user interaction with the text field.
    @Published var disabled: Bool

    /// A custom view to display on the left side of the text field.
    @Published var leftView: AnyView?

    /// A custom view to display on the right side of the text field.
    @Published var rightView: AnyView?

    // MARK: Initialization

    init(
        isSecureTextEntry: Bool = false,
        titleColor: Color = .gray,
        placeholderColor: Color = .gray,
        placeholderFont: Font = .system(size: 15),
        font: Font = .system(size: 15),
        borderWidth: CGFloat = .zero,
        borderColor: Color = .gray,
        focusedBorderColor: Color = .gray,
        cornerRadius: CGFloat = .zero,
        placeholderBottomPadding: CGFloat = .zero,
        disabled: Bool = false,
        leftView: AnyView? = nil,
        rightView: AnyView? = nil
    ) {
        self.isSecureTextEntry = isSecureTextEntry
        self.titleColor = titleColor
        self.placeholderColor = placeholderColor
        self.placeholderFont = placeholderFont
        self.font = font
        self.borderWidth = borderWidth
        self.borderColor = borderColor
        self.focusedBorderColor = focusedBorderColor
        self.cornerRadius = cornerRadius
        self.placeholderBottomPadding = placeholderBottomPadding
        self.disabled = disabled
        self.leftView = leftView
        self.rightView = rightView
    }
}


Sources/FloatingTextField/Classes/UI/FloatingTextField.swift
//
// floating-text-field
// Copyright © 2023 Space Code. All rights reserved.
//

import SwiftUI

// MARK: - FloatingTextField

/// A control that displays an editable text interface.
///
/// You create a text field with a label and a binding to a value. The text field updates this value continuously as
/// the user types or otherwise edits the text in the field.
///
/// The following example demonstrates a text field designed for accepting a username with custom configuration.
///
/// ```swift
/// struct ContentView: View {
///
///    @State private var username: String = ""
///
///    var body: some View {
///        FloatingTextField($username, placeholder: "title")
///            .borderWidth(1)
///            .borderCornerRadius(12)
///            .borderColor(Color.black)
///            .textFont(Font.system(size: 17.0))
///            .placeholderFont(Font.system(size: 14.0))
///    }
/// }
/// ```
public struct FloatingTextField: View {
    // MARK: Properties

    /// The top placeholder size.
    @State private var placeholderSize: CGSize = .zero
    /// The text field size.
    @State private var textFieldSize: CGSize = .zero
    /// The left view size.
    @State private var leftViewSize: CGSize = .zero

    /// When focus moves to the view, the binding sets the bound value to true.
    @State private var isSelected = false
    /// The text to display and edit.
    @Binding private var text: String
    /// When focus moves to the view, the binding sets the bound value to true.
    @FocusState private var isFocused: Bool
    /// The text field configuratuion.
    @ObservedObject var configuration = FloatingTextFieldConfiguration()

    /// A bool value that indicates whether the placeholder is shown or not.
    private var isTopPlaceholderVisible: Bool {
        !text.isEmpty || isSelected
    }

    private var leftSideOffset: CGFloat {
        configuration.leftView != nil ? leftViewSize.width + .spacing : .zero
    }

    /// The title of the text view, describing its purpose.
    private let placeholder: String

    // MARK: Initialization

    /// Creates a text field with a text label generated and a placeholder.
    ///
    /// - Parameters:
    ///   - text: The text to display and edit.
    ///   - placeholder: The title of the text view, describing its purpose.
    public init(_ text: Binding<String>, placeholder: String) {
        _text = text
        self.placeholder = placeholder
    }

    // MARK: View

    public var body: some View {
        GeometryReader { geometry in
            VStack(alignment: .leading, spacing: .zero) {
                ZStack(alignment: .leading) {
                    topPlaceholderView
                        .background(BackgroundGeometryReader(size: $placeholderSize))
                        .opacity(isTopPlaceholderVisible ? 1 : .zero)
                        .offset(x: leftSideOffset, y: isTopPlaceholderVisible ? -offset(for: placeholderSize) : .zero)

                    HStack(alignment: .center, spacing: .spacing) {
                        leftView
                            .background(BackgroundGeometryReader(size: $leftViewSize))
                        contentView
                            .background(BackgroundGeometryReader(size: $textFieldSize))
                            .offset(y: isTopPlaceholderVisible ? offset(for: textFieldSize) : .zero)
                        rightView
                    }
                }
            }
            .onChange(of: isFocused) { newValue in
                withAnimation(.easeInOut(duration: .animationDuration)) {
                    isSelected = newValue
                }
            }
            .padding()
            .frame(maxHeight: geometry.size.height)
            .overlay(
                RoundedRectangle(cornerRadius: configuration.cornerRadius)
                    .strokeBorder(
                        isSelected ? configuration.borderColor : configuration.focusedBorderColor,
                        lineWidth: configuration.borderWidth
                    )
            )
        }
    }

    // MARK: Private

    private var topPlaceholderView: some View {
        Text(placeholder)
            .font(configuration.placeholderFont)
            .foregroundColor(configuration.placeholderColor)
            .frame(alignment: .leading)
    }

    private var placeholderView: some View {
        Text(placeholder)
            .foregroundColor(configuration.placeholderColor)
            .font(configuration.font)
    }

    private var secureField: some View {
        SecureField(String.empty, text: $text.animation())
            .foregroundColor(configuration.titleColor)
            .multilineTextAlignment(.leading)
            .font(configuration.font)
            .focused($isFocused)
            .disabled(configuration.disabled)
    }

    private var textField: some View {
        TextField(String.empty, text: $text.animation())
            .foregroundColor(configuration.titleColor)
            .multilineTextAlignment(.leading)
            .font(configuration.font)
            .focused($isFocused)
            .disabled(configuration.disabled)
    }

    private var contentView: some View {
        ZStack(alignment: .leading) {
            if !isTopPlaceholderVisible {
                placeholderView
            }

            if configuration.isSecureTextEntry {
                secureField
            } else {
                textField
            }
        }
    }

    private var leftView: some View {
        configuration.leftView.map { $0 }
    }

    private var rightView: some View {
        configuration.rightView.map { $0 }
    }

    // MARK: Helpers

    private func offset(for size: CGSize) -> CGFloat {
        (size.height + configuration.placeholderBottomPadding) / 2.0
    }
}

// MARK: - Constants

private extension String {
    static let empty = ""
}

private extension TimeInterval {
    static let animationDuration = 0.178
}

private extension CGFloat {
    static let spacing: CGFloat = 8.0
}

// MARK: - Previews

#if DEBUG
    // swiftlint:disable:next type_name
    struct FloatingTextField_Preview: PreviewProvider {
        static var previews: some View {
            ForEach(ColorScheme.allCases, id: \.self) {
                FloatingTextField(.constant(""), placeholder: "placeholder")
                    .borderWidth(1)
                    .cornerRadius(12)
                    .frame(height: 60.0)
                    .preferredColorScheme($0)

                FloatingTextField(.constant("text"), placeholder: "placeholder")
                    .borderWidth(1)
                    .cornerRadius(12)
                    .focusedBorderColor(.gray)
                    .frame(height: 60.0)
                    .preferredColorScheme($0)
            }
        }
    }
#endif
